homeassistant/components/geocaching/sensor.py

```diff
@@ -1,74 +1,132 @@
+            icon="mdi:notebook-edit-outline",
+            default_enabled=True,
+        ),
+        "hide_count": GeocachingSensorSettings(
+            name="Total hides",
+            section="user",
+            state="hide_count",
+            unit_of_measurement="caches",
+            device_class=None,
+            icon="mdi:eye-off-outline",
+            default_enabled=True,
+        ),
+        "favorite_points": GeocachingSensorSettings(
+            name="Favorite points",
+            section="user",
+            state="favorite_points",
+            unit_of_measurement="points",
+            device_class=None,
+            icon="mdi:heart-outline",
+            default_enabled=True,
+        ),
+        "souvenir_count": GeocachingSensorSettings(
+            name="Total souvenirs",
+            section="user",
+            state="souvenir_count",
+            unit_of_measurement="souvenirs",
+            device_class=None,
+            icon="mdi:license",
+            default_enabled=True,
+        ),
+        "awarded_favorite_points": GeocachingSensorSettings(
+            name="Awarded favorite points",
+            section="user",
+            state="awarded_favorite_points",
+            unit_of_measurement="points",
+            device_class=None,
+            icon="mdi:heart",
+            default_enabled=True,
+        ),
+    }
+
     coordinator = hass.data[DOMAIN][entry.entry_id]["COORD"]
     async_add_entities(
-        [GeocachingSensor(coordinator, key=key) for key in SENSOR_ENTITIES]
+        [
+            GeocachingSensor(coordinator, key=key, settings=sensor_data[key])
+            for key in sensor_data
+        ]
     )
 
 
 class GeocachingSensor(GeocachingEntity, SensorEntity):
     """Representation of a Sensor."""
 
+    settings: GeocachingSensorSettings
+    key: str
+
     def __init__(
-        self, coordinator: GeocachingDataUpdateCoordinator, *, key: str
+        self,
+        coordinator: GeocachingDataUpdateCoordinator,
+        *,
+        key: str,
+        settings: GeocachingSensorSettings,
     ) -> None:
         """Initialize the Geocaching sensor."""
-        self.key = key
 
         super().__init__(
             coordinator,
-            enabled_default=bool(SENSOR_ENTITIES[key][ATTR_DEFAULT_ENABLED]),
-            icon=str(SENSOR_ENTITIES[key][ATTR_ICON]),
-            name=f"Geocaching {coordinator.data.user.reference_code} {SENSOR_ENTITIES[key][ATTR_NAME]}",
+            enabled_default=settings["default_enabled"],
+            icon=settings["icon"],
+            name=f"Geocaching {coordinator.data.user.reference_code} {settings['name']}",
         )
+        self.settings = settings
+        self.key = key
 
     @property
     def unique_id(self) -> str:
         """Return the unique ID for this sensor."""
         return f"geocaching_{self.coordinator.data.user.reference_code}_{self.key}"
 
     @property
-    def state(self):
+    def state(self) -> Any:
```

Comment:
There is `StateType` in `homeassistant.helpers.typing`, you can use it. Especially if you have any idea what `getattr(section, self.settings["state"])` may return.

Reply:
I don't know what type getattr returns, That depends on the sctual sensor. That prevent me from using StateType I presume?

Reply:
> `coordinator.data` is `GeocachingStatus` which only has `user` and `trackables` fields. Does it also have some fields which are created during runtime?

The sensor is pulling data from the GeocachingStatus object properties using the section and state values from the GeocachingSensorSettings. They can be of different types, and the actual type is therefore only available at runtime.

> Btw, `geocachingapi` library seems to be covered by type annotations, but you don't check it and don't export type information from it. This can be improved.

Do you have time to elaborate on this, with an example or something? As mentioned before I am not a python developer, so maybe you can help  me forward with that :). Thanks in advance!

Reply:
> The sensor is pulling data from the GeocachingStatus object properties using the section and state values from the GeocachingSensorSettings. They can be of different types, and the actual type is therefore only available at runtime.

Can this API be improved? Creating new attributes in runtime is not the best practice. It could be some dict with data in `GeocachingStatus`.

> Do you have time to elaborate on this, with an example or something?
1. Create Github Action which runs some checks: black, isort, mypy. Maybe also pylint and flake8. Type hints which are not checked are not very useful.
2. To export typing information with the package you need to create `py.typed` marker (empty file) and include it in the package root.

This is a good example how to setup CI with Github Actions and p`re-commit` and manage dependencies using poetry (I prefer it much more compare to setuptools): https://github.com/leikoilja/glocaltokens

Feel free to ask more questions if something is not clear.

Reply:
> Can this API be improved? Creating new attributes in runtime is not the best practice. It could be some dict with data in GeocachingStatus.

Its not creating attributes but looking up the requested in the typed result from the API. However, I agree there's some room for improvement there. I will add it to my todo list for a later, more extended, version of the integration. 

I will add the typing enhancement of the python package to  my todo list for the next version of the API. Thank you for your good suggestions!